Update translations handling documentation in Docs (#560)
sharetribe/flex-docs: src/docs/ftw-content/hosted-translations/index.md

---
title: Hosted translations with Asset Delivery API
slug: hosted-translations
updated: 2022-05-13
category: ftw-content
ingress:
  This article describes how hosted translations work in Flex Template
  for Web (FTW).
published: true
---

Starting from 2022-05, FTW template translations can be managed both in
the built-in translation files and in Flex Console. This article
describes how the FTW template uses the hosted translations and merges
them with the built-in translations

_**Note:** If you want to implement this feature into your pre-version

> XX.XX FTW template, you can see the necessary modifications in the PRs

Reviewer: Related to this, I think the next version are:
- FTW-daily v8.5
- FTW-hourly v10.5
- FTW-product: v9.2

aka minor version update

Author: Thanks, added this!

> for [ftw-daily](), [ftw-hourly](), and [ftw-product]() todo: links. Read

Reviewer: I'd use just the FTW-daily. Other PR:s are coming only after FTW-daily has had release (update is taken from release tag)
https://github.com/sharetribe/ftw-daily/pull/1510

Author: Fixed!

more:_

- _[Translations in Flex Console](/concepts/translations/)_
- _[How built-in translations work in the FTW templates](ftw/how-to-change-ftw-bundled-ui-translations/)_

## Hosted translations

**Hosted translations** are translations that the marketplace operator
can edit in Flex Console. The client app then needs to fetch them using
the Asset Delivery API.

It is good to note that even if the operator adds some hosted
translations using Flex Console, the FTW template still needs to have a
built-in translation file for the translation keys that do not have a
value in the hosted asset. That way, the UI can still render something
meaningful for the parts of the page that the operator has not modified.

There are two ways to fetch translation assets using Asset Delivery API:
by version or by alias.

#### Fetching translations by version

All assets are identifiable by their version, and versions are
immutable. Therefore if you fetch assets by version, they can be cached

> for an extended period of time. Read more about [caching assets]() TODO:

Reviewer: TODO
I'm not sure if this is meant to point to FTW-related article or somewhere else?

Reviewer: maybe `/references/assets/#asset-versioning`

Reviewer: or 

`https://www.sharetribe.com/api-reference/asset-delivery-api.html#caching`

LINK. When fetching translations by version, they are cached for an

> extended period of time, which helps to avoid unnecessary data loading.

Reviewer: Maybe one more sencence?
"Since Asset Delivery API sets ***Cache-Control*** header for these responses, the browser knows to cache these responses on its own."

Since FTW templates use Flex SDK, the response always contains the data
for the requested asset, even if the asset has not changed in the
specified version. However, when using the API endpoint without SDK, it
is important to take into account the API's [asset redirect behavior]()

> TODO: LINK.

Reviewer: TODO

Reviewer: This is a bit confusing paragraph. Maybe there should be something before it so that the story is a bit more clear:

```markdown
Hosted assets are versioned as a [whole asset tree](/references/assets/#asset-versioning) - a bit similar to how Git works. 
Individual asset files might have not changed when the whole version has changed and this might cause
[HTTP redirects](https://www.sharetribe.com/api-reference/asset-delivery-api.html#http-redirects)

Since FTW templates use Flex SDK, the response always contains the data
for the requested asset, even if the asset has not changed in the
specified version. However, when using the API endpoint without SDK, it
is important to take into account the API's [asset redirect behavior](https://www.sharetribe.com/api-reference/asset-delivery-api.html#http-redirects).
```

```js
sdk.assetByVersion({
  path: 'content/translations.json',
  version: '<some-hash-string>',
});
```

#### Fetching translations by alias

In addition to fetching assets by version, you can fetch them by a
specific alias instead of a version. Currently, translations can be

> fetched with the alias `latest`, which returns the most recently updated

Reviewer: ```suggestion
fetched with the alias _"latest"_, which returns the most recently updated
```
This is just a preference, but we have had too many uses for code-block highlights in the past. Since this is not code. Personally, I'd prefer to use bolds or italics instead.

version of the translations. The response also contains the version
information for the most recent asset, so that subsequent fetches can be
done based on asset version.

> When fetching by alias, the cache time is 5 seconds for development

Reviewer: ```suggestion
When fetching by alias, the cache time is a few seconds for development
```

Currently, it's 5 seconds, but API has reserved some leeway: "up to 10 seconds"

Reviewer: Although, the paragraph ends with "These cache times are subject to change." So, maybe this is fine?

environment and up to 5 minutes for production environment. In other
words, it can take up to 5 minutes for translation updates to be visible
in a production environment. These cache times are subject to change.

```js
sdk.assetByAlias({
  path: 'content/translations.json',
  alias: 'latest',
});
```

We have added a new global Redux file
(_src/ducks/hostedAssets.duck.js_), which exports Redux Thunk function
called _fetchAppAssets_. This function actually makes the calls to the
Asset Delivery API.

Translations are part of app-wide hosted assets that have configuration
in _src/config.js_

```js
// CDN assets for the app. Configurable through Flex Console.
// Currently, only translation.json is available.
const appCdnAssets = {
  translations: 'content/translations.json',
};

> ```

Reviewer: This is used by both SDK calls (alias and versioned), so, it's probably better to tell this already before these sub headings.

----

**Hosted translations** are translations that the marketplace operator
can edit in Flex Console. The client app then needs to fetch them using
the Asset Delivery API.

It is good to note that even if the operator adds some hosted
translations using Flex Console, the FTW template still needs to have a
built-in translation file for the translation keys that do not have a
value in the hosted asset. That way, the UI can still render something
meaningful for the parts of the page that the operator has not modified.

FTW templates have specified hosted translations as part of the app-wide configuration
in _src/config.js_.

```js
// CDN assets for the app. Configurable through Flex Console.
// Currently, only translation.json is available.
// Note: the path must match the path defined in Asset Delivery API
const appCdnAssets = {
  translations: 'content/translations.json',
};
```

In addition, they have added a new global Redux file
(_src/ducks/hostedAssets.duck.js_), which exports the Redux Thunk function
called _fetchAppAssets_. This function actually makes the calls to the
Asset Delivery API.

There are two ways to fetch translation assets using Asset Delivery API:
by the version or by alias.

## How production build works with hosted translations

This setup is in use if you run **_yarn start_** in your host
environment or **_yarn run dev-server_** on your local machine.

1. Export **fetchAppAssets** from _src/index.js_ to make it available
   for the server.
2. SSR: initialize the store

   ```shell
   └── server
       └── dataLoader.js
   ```

   ```js
   let translations = {};
   const store = configureStore({}, sdk);
   return store.dispatch(fetchAppAssets(config.appCdnAssets));
   ```

3. SSR: **fetchAppAssets** thunk fetches the latest version of

>    `content/translations.json` asset by using `latest` alias with

Reviewer: No need to change, but a personal preference to reduce the overuse of "code" -highlights that are pretty strong when you try to read an article. 

```suggestion
   _"content/translations.json"_ asset by using ***latest*** alias with
```

   `sdk.assetByAlias`
   - This ensures that the server-side rendering has the most recent
     version of asset to render the page
4. SSR: make a _loadData_ call if the route specifies it and

   ```js
     .then(fetchedAssets => {
       translations = fetchedAssets?.translations?.data || {};
       return Promise.all(dataLoadingCalls);
   })
   ```

5. SSR: asset **version** is saved to the store and passed to the
   browser through the _"preloadedState"_.

   - Server-side rendering must match with client-side rendering when
     browser hydrates the server-side rendered content.
   - To ensure that client-side rendering has the same version of
     translations, the asset version is passed (through preloaded state)
     to front end.
   - Note: FTW does not pass the translations themselves from the server
     to browser. The reason is that if browser fetches the versioned
     asset file directly, it can leverage browser's cache. So, every

>      page load, after the initial one, will use the `translation.json`

Reviewer: ```suggestion
     page load, after the initial one, will use the ***translation.json***
```

        messages={{ ...localeMessages, ...hostedTranslations }}
        textComponent="span"
      >
    {/* etc. */}
  ```

7. Browser: initialize the store with a preloaded state
   - The asset version that the SSR used, is included in that preloaded
     state
8. Browser: **fetchAppAssets** thunk fetch assets using asset version:
   calls `sdk.assetByVersion`.
9. Browser: make _loadData_ call
10. Browser: hydrate the `<ClientApp>` and pass translations as props.

- Hosted translations are merged with default translations in
  `ClientApp` component.
  ```js
  <ClientApp store={store} hostedTranslations={translations} />
  ```

## How development build works with hosted translations

This is setup is in use if you run **_yarn run dev_** on local machine.

1. Browser: initialize the store
2. Browser: **fetchAppAssets** thunk fetches assets using 'latest'
   alias: calls `sdk.assetByAlias`
   - Because SSR is not available to fetch the latest version of the
     asset files, this call needs to be made from browser.
3. Browser: make _loadData_ call
4. Browser: render the `<ClientApp>` and pass translations as props.

## Read more

If you want to read more, here are some pointers:

> - [Asset Delivery API](/references/<the-correct-link-here>/)

Author: TODO

> - [Short intro to SSR](https://deploy-preview-565--sharetribe-flex-docs-site.netlify.app/docs/ftw/how-routing-works-in-ftw/#a-brief-introduction-to-ssr)

Reviewer 2: Should this be pointing to a deploy preview?

Author: Fixed to an internal link
